### PR TITLE
Add rabbitmqClusterReference.connectionSecret

### LINKED ISSUE #135: Be able to use existing RabbitMQ clusters that are not created/managed by https://github.com/rabbitmq/cluster-operator
**Is your feature request related to a problem? Please describe.**
Some users might want to decouple the creation of a rabbitmqcluster from using messaging-topology operator to manage resources on it. Think of it as "Bring my own RabbitMQ". Currently I do not see a way to do this, since all the existing resources require a reference to RabbitMQCluster, for example here:
https://github.com/rabbitmq/messaging-topology-operator/blob/main/api/v1beta1/binding_types.go#L43

And seems like we don't really need the rabbitmqcluster reference except to pull in the connection information so that we can create the client connection, so wondering if we could investigate providing a mechanism to be able to specify the connection information (say in a secret with predefined fields for example) so that we could support the use case for bringing your own existing RabbitMQ broker and not require that it be created via the RabbitMQCluster operator.
https://github.com/rabbitmq/messaging-topology-operator/blob/main/internal/rabbitmq_client_factory.go#L76

**Describe the solution you'd like**
Perhaps instead of requiring that the credentials for the connection come from RabbitMQCluster reference:
https://github.com/rabbitmq/messaging-topology-operator/blob/main/api/v1beta1/binding_types.go#L43

We could add another way to specify required connection information to be able to connect to existing RabbitMQ clusters?

**Describe alternatives you've considered**
I don't currently see a way to support this use case because you must use a RabbitMQClusterReference.

**Additional context**

### PR DESCRIPTION
**Pull request by ChunyiLyu**
This closes #135 

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed
**Note to contributors:** remember to re-generate client set if there are any API changes

## Summary Of Changes

Will be squashed merged due to messy history.
Needs to create a followup for handling TLS connection

### CODE REVIEW

File: api/v1beta1/exchange_webhook.go [hunk 1]
```diff
@@ -19,8 +19,20 @@ func (r *Exchange) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
+			field.Forbidden(field.NewPath("spec", "rabbitmqClusterReference"),
+				"do not provide both spec.rabbitmqClusterReference.name and spec.rabbitmqClusterReference.connectionSecret"))
+	}
+
+	if e.Spec.RabbitmqClusterReference.Name == "" && e.Spec.RabbitmqClusterReference.ConnectionSecret == nil {
+		return apierrors.NewForbidden(e.GroupResource(), e.Name,
+			field.Forbidden(field.NewPath("spec", "rabbitmqClusterReference"),
+				"must provide either spec.rabbitmqClusterReference.name or spec.rabbitmqClusterReference.connectionSecret"))
+	}
 	return nil
```
Comment (coro) [inline api/v1beta1/exchange_webhook.go:36]: Could we extract this test case into a common function used by all of the webhooks?

File: api/v1beta1/rabbitmq_cluster_reference.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	// Have to set either name or connectionSecret, but not both.
+	// +kubebuilder:validation:Optional
+	Name string `json:"name,omitempty"`
+	// The namespace of the RabbitMQ cluster to reference.
+	// Defaults to the namespace of the requested resource if omitted.
+	// +kubebuilder:validation:Optional
+	Namespace string `json:"namespace,omitempty"`
+	// Secret contains the http management uri for the RabbitMQ cluster.
+	// The Secret must contain the key `uri` or operator will error.
+	// Have to set either name or connectionSecret, but not both.
```
Comment (coro) [inline api/v1beta1/rabbitmq_cluster_reference.go:16]: Could do with mentioning the username/password keys too

File: internal/rabbitmq_client_factory.go [hunk 3]
```diff
@@ -54,21 +54,27 @@ var RabbitholeClientFactory RabbitMQClientFactory = func(rmq *rabbitmqv1beta1.Ra
 	return generateRabbitholeClient(rmq, svc, credsProvider, hostname, certPool)
 }
 
-// returns a http client for the given RabbitmqCluster
+// generateRabbitholeClient returns a http client for the given RabbitmqCluster
+// if provided RabbitmqCluster is nil, generateRabbitholeClient uses username, passwords, and uri
+// information from credsProvider to generate a rabbit client
 func generateRabbitholeClient(rmq *rabbitmqv1beta1.RabbitmqCluster, svc *corev1.Service, credsProvider CredentialsProvider, hostname string, certPool *x509.CertPool) (rabbitmqClient RabbitMQClient, err error) {
```
Comment (coro) [inline internal/rabbitmq_client_factory.go:60]: I'm not sure I like the interface this function now provides. Seems like in order for code to call this function, it has to be tightly coupled to how it behaves under the covers. In this case, consuming code must know that rmq and svc can be nil, and that using the data that is provided in credsProvider depends on whether rmq is nil or not. It feels like the function is doing too much at once as a result.

I would try to make the factory functions as small as possible, and only take in what they need to construct the client. I would personally either move the logic for constructing the inputs to the factory either to the controllers, its own library, or separate functions in this library.

**Review by coro — COMMENTED** [review]
Will test live now.

**Review by coro — COMMENTED** [review]
One neat thing this provides is that it solves this issue: https://github.com/rabbitmq/messaging-topology-operator/issues/287

**Comment by Zerpet** [comment]
> I'm seeing an issue where if I create a topology object, then the RabbitMQ cluster disappears for whatever reason, the object is stuck in failed delete forever as the operator hits tcp timeouts trying to connect.
> 
> In the case where we're using a LocalClusterReference, and we fail to find the RMQ, we assume it's been deleted and we delete the underlying topology object; I'd like to see a similar thing here potentially. There is a trade off, in that a cluster may only be temporarily unavailable while a user does a delete operation, and we'd leave some state around, but I think that's a better solution than having users do a `k delete --force`.

This behaviour is similar to what's described in #277. Big plus if this PR also fixes that problem.

File: docs/examples/rmqsourcecrd [hunk 4]
```diff
@@ -0,0 +1,130 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (coro) [inline docs/examples/rmqsourcecrd:1]: What is this file? Accidental checkin?

**Review by coro — APPROVED** [review]
Following on from our chat on Zoom, LGTM! Just one concern about an accidentally included file potentially?

File: docs/examples/rmqsourcecrd (hunk 4 above)
Comment (ChunyiLyu) [reply docs/examples/rmqsourcecrd]: removed

**Pull request merged**